pynucastro/rates/rate.py

```diff
@@ -1635,6 +1636,76 @@ def eval(self, T, rhoY=None):
             return r*z_r/z_p
         return r
 
+    def function_string_py(self):
+        """
+        Return a string containing python function that computes the
+        rate
+        """
+
+        fstring = ""
+        fstring += "@numba.njit()\n"
+        fstring += f"def {self.fname}(tf):\n"
+        fstring += f"    # {self.rid}\n"
+        fstring += "    rate = 0.0\n\n"
+
+        for s in self.sets:
+            fstring += f"    # {s.labelprops[0:5]}\n"
+            set_string = s.set_string_py(prefix="rate", plus_equal=True)
+            for t in set_string.split("\n"):
+                fstring += "    " + t + "\n"
+
+        if self.use_pf:
+
+            fstring += "\n"
+            for nucr in self.rate.reactants:
+                fstring += f"    {nucr}_temp_array = np.array({list(nucr.partition_function.temperature/1.0e9)})\n"
+                fstring += f"    {nucr}_pf_array = np.array({list(nucr.partition_function.partition_function)})\n"
+                fstring += f"    {nucr}_pf_exponent = np.interp(tf.T9, xp={nucr}_temp_array, fp=np.log10({nucr}_pf_array))\n"
+                fstring += f"    {nucr}_pf = 10.0**{nucr}_pf_exponent\n"
+                fstring += "\n"
+            for nucp in self.rate.products:
+                fstring += f"    {nucp}_temp_array = np.array({list(nucp.partition_function.temperature/1.0e9)})\n"
+                fstring += f"    {nucp}_pf_array = np.array({list(nucp.partition_function.partition_function)})\n"
+                fstring += f"    {nucp}_pf_exponent = np.interp(tf.T9, xp={nucp}_temp_array, fp=np.log10({nucp}_pf_array))\n"
+                fstring += f"    {nucp}_pf = 10.0**{nucp}_pf_exponent\n"
+                fstring += "\n"
+
+            fstring += "\n"
+            fstring += "    "
+            fstring += "z_r = "
+            fstring += "*".join([f"{nucr}_pf" for nucr in self.rate.reactants])
+
+            fstring += "\n"
+            fstring += "    "
+            fstring += "z_p = "
+            for index, nucp in enumerate(self.rate.products):
+                if index == len(self.rate.products)-1:
+                    fstring += f"{nucp}_pf"
+                    break
+                fstring += f"{nucp}_pf*"
+
+            fstring += "\n"
+            fstring += "    rate *= z_r/z_p\n"
+
+        fstring += "\n"
+        fstring += "    return rate\n\n"
+        return fstring
+
+    def counter_factors(self):
+
+        react_counts = Counter(self.rate.reactants)
+        prod_counts = Counter(self.rate.products)
```

Comment:
We should assert that the reactants and products don't overlap

Reply:
the point is that if they do, it will still be ok by using two different counters.

Reply:
Sounds good

Reply:
I was thinking in the following:
1. Take the counter of reactants and products.
2.  For each nuc in reactant+product, multiply or divide by N_a,
3.  Reduce the nuc counter by 1